### ISSUE TITLE
bgfx hlsl screen chain crashes with high resolution games 

### ISSUE THREAD

**Issue by belegdol** [mamedev/mame#5857]
It seems like the hlsl issue might be separate. I have built latest git master under windows 10 and found the following:

renderer |  ikaruga default | ikaruga hlsl | firefox default | firefox hlsl
--- | --- | --- | --- | ---
dx9 | crash | crash | crash | crash
dx11 | crash | crash | crash | crash
dx12 | crash | crash | runs | crash
opengl | runs | crash | crash | crash
vulkan | runs | crash | crash | crash

In summary: hlsl crashes across the board while the default screen chain works on opengl and vulkan for ikaruga and on dx12 for firefox.

_Originally posted by @belegdol in https://github.com/mamedev/mame/issues/5830#issuecomment-548761377_
It appears that the hlsl crash is related to game resolution. firefox, ikaruga and n64 all crash whereas vf2, mk4 and umk3 do not.

**Comment by belegdol** [comment]
This is the backtrace of the crash under linux:

In all 4 cases (firefox and ikaruga with opengl and vulkan) the backtraces for hlsl look very similar:
```
Thread 1 "mame64" received signal SIGSEGV, Segmentation fault.
0x00000000006d87eb in chain_manager::update_screen_textures (this=0x29d77c0, view=<optimized out>, starting_prim=<optimized out>, 
    window=...) at ../../../../../src/osd/modules/render/bgfx/chainmanager.cpp:515
515		return m_screen_prims.size();
(gdb) bt
#0  0x00000000006d87eb in chain_manager::update_screen_textures (this=0x29d77c0, view=<optimized out>, 
    starting_prim=<optimized out>, window=...) at ../../../../../src/osd/modules/render/bgfx/chainmanager.cpp:515
#1  0x0000000000674f4d in renderer_bgfx::draw (this=0x2d0c7c0, update=<optimized out>) at ../../../../../src/emu/render.h:360
#2  0x000000000064d03a in sdl_window_info::update (this=0x29a3f90) at /usr/include/c++/9/bits/unique_ptr.h:352
#3  sdl_window_info::update (this=0x29a3f90) at ../../../../../src/osd/sdl/window.cpp:560
#4  0x000000000064a6dd in sdl_osd_interface::update (this=0x7fffffffcfa0, skip_redraw=<optimized out>)
    at ../../../../../src/osd/sdl/video.cpp:108
#5  0x0000000001b3919b in video_manager::frame_update (this=0x43599c0, from_debugger=from_debugger@entry=false)
    at ../../../../../src/emu/video.h:54
#6  0x0000000001afaade in screen_device::vblank_begin (this=0x2c66d30) at /usr/include/c++/9/bits/unique_ptr.h:352
#7  0x0000000001afd8b5 in screen_device::device_timer (this=0x2c66d30, timer=..., id=<optimized out>, param=<optimized out>, 
    ptr=<optimized out>) at ../../../../../src/emu/screen.cpp:897
#8  0x0000000001af4942 in device_t::timer_expired (ptr=<optimized out>, param=<optimized out>, id=<optimized out>, timer=..., 
    this=<optimized out>) at ../../../../../src/emu/device.h:609
#9  device_scheduler::execute_timers (this=<optimized out>) at ../../../../../src/emu/schedule.cpp:894
#10 device_scheduler::timeslice (this=this@entry=0x7fffffffc298) at ../../../../../src/emu/schedule.cpp:534
#11 0x0000000001a9e568 in running_machine::run (this=this@entry=0x7fffffff6d20, quiet=quiet@entry=false)
    at ../../../../../src/emu/machine.cpp:367
#12 0x0000000000732a25 in mame_machine_manager::execute (this=this@entry=0x29de220)
    at ../../../../../src/frontend/mame/mame.cpp:261
#13 0x00000000007ce73a in cli_frontend::start_execution (this=0x7fffffffcd90, manager=0x29de220, args=...)
    at ../../../../../src/frontend/mame/clifront.cpp:264
#14 0x00000000007ce91e in cli_frontend::execute (this=this@entry=0x7fffffffcd90, 
    args=std::vector of length 12, capacity 12 = {...}) at ../../../../../src/frontend/mame/clifront.cpp:280
#15 0x0000000000730cd8 in emulator_info::start_frontend (options=..., osd=..., args=std::vector of length 12, capacity 12 = {...})
    at ../../../../../src/frontend/mame/mame.cpp:392
#16 0x00000000004df9cb in main (argc=<optimized out>, argv=<optimized out>) at ../../../../../src/osd/sdl/sdlmain.cpp:217
```

_Originally posted by @belegdol in https://github.com/mamedev/mame/issues/5830#issuecomment-548718916_

**Comment by belegdol** [comment]
With ikaruga and n64, I can launch the game with screen chain set to none and switch to hlsl without crashes. The crash only happens when hlsl is passed from command-line or from the ini.
With firefox attempting to switch the chain kills the sound and makes everyting go green, but this is probably an unrelated problem.

**Comment by belegdol** [comment]
I have been investigating the resolution issue further: tgmj also crashes with hlsl, but only after showing the test colour bars - with almost the same backtrace. This is with opengl bgfx backend:

```
(gdb) run -verbose tgmj
Starting program: /usr/bin/mame -verbose tgmj
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Attempting load of mame.ini
Parsing mame.ini
Attempting load of mame.ini
Parsing mame.ini
Attempting load of horizont.ini
Attempting load of arcade.ini
Attempting load of raster.ini
Attempting load of source/zn.ini
Attempting load of coh3002c.ini
Attempting load of tgmj.ini
Starting plugin data...
Attempting load of mame.ini
Parsing mame.ini
Attempting load of mame.ini
Parsing mame.ini
Attempting load of horizont.ini
Attempting load of arcade.ini
Attempting load of raster.ini
Attempting load of source/zn.ini
Attempting load of coh3002c.ini
Attempting load of tgmj.ini
Available videodrivers: x11 wayland KMSDRM dummy 
Current Videodriver: x11
	Display #0
		Renderdrivers:
			    opengl (0x0)
			 opengles2 (0x0)
			  software (0x0)
Available audio drivers: 
	pulseaudio          
	alsa                
	dsp                 
	jack                
	disk                
	dummy               
Build version:      0.215 (unknown)
Build architecure:  
Build defines 1:    SDLMAME_UNIX=1 SDLMAME_X11=1 SDLMAME_LINUX=1 
Build defines 1:    LSB_FIRST=1 PTR64=1 
SDL/OpenGL defines: SDL_COMPILEDVERSION=2010 USE_OPENGL=1 
Compiler defines A: __GNUC__=9 __GNUC_MINOR__=2 __GNUC_PATCHLEVEL__=1 __VERSION__="9.2.1 20190827 (Red Hat 9.2.1-1)" 
Compiler defines B: __amd64__=1 __x86_64__=1 __unix__=1 
Compiler defines C: _FORTIFY_SOURCE=2 __USE_FORTIFY_LEVEL=2 
Enter init_monitors
Adding monitor screen0 (1920 x 1080)
Leave init_monitors
Enter sdlwindow_init

Hints:
	SDL_FRAMEBUFFER_ACCELERATION             (NULL)
	SDL_RENDER_DRIVER                        (NULL)
	SDL_RENDER_OPENGL_SHADERS                (NULL)
	SDL_RENDER_SCALE_QUALITY                 (NULL)
	SDL_RENDER_VSYNC                         (NULL)
	SDL_VIDEO_X11_XVIDMODE                   (NULL)
	SDL_VIDEO_X11_XINERAMA                   (NULL)
	SDL_VIDEO_X11_XRANDR                     (NULL)
	SDL_GRAB_KEYBOARD                        (NULL)
	SDL_VIDEO_MINIMIZE_ON_FOCUS_LOSS         (NULL)
	SDL_IOS_IDLE_TIMER_DISABLED              (NULL)
	SDL_IOS_ORIENTATIONS                     (NULL)
	SDL_XINPUT_ENABLED                       (NULL)
	SDL_GAMECONTROLLERCONFIG                 (NULL)
	SDL_JOYSTICK_ALLOW_BACKGROUND_EVENTS     (NULL)
	SDL_ALLOW_TOPMOST                        (NULL)
	SDL_TIMER_RESOLUTION                     (NULL)
	SDL_RENDER_DIRECT3D_THREADSAFE           (NULL)
	SDL_VIDEO_ALLOW_SCREENSAVER              (NULL)
	SDL_ACCELEROMETER_AS_JOYSTICK            (NULL)
	SDL_MAC_CTRL_CLICK_EMULATE_RIGHT_CLICK   (NULL)
	SDL_VIDEO_WIN_D3DCOMPILER                (NULL)
	SDL_VIDEO_WINDOW_SHARE_PIXEL_FORMAT      (NULL)
	SDL_VIDEO_MAC_FULLSCREEN_SPACES          (NULL)
	SDL_MOUSE_RELATIVE_MODE_WARP             (NULL)
	SDL_RENDER_DIRECT3D11_DEBUG              (NULL)
	SDL_VIDEO_HIGHDPI_DISABLED               (NULL)
	SDL_WINRT_PRIVACY_POLICY_URL             (NULL)
	SDL_WINRT_PRIVACY_POLICY_LABEL           (NULL)
	SDL_WINRT_HANDLE_BACK_BUTTON             (NULL)
Leave sdlwindow_init
Enter sdl_info::create
[New Thread 0x7fffe0f1d700 (LWP 107595)]
[New Thread 0x7fffc9a37700 (LWP 107598)]
Audio: Start initialization
Audio: Driver is pulseaudio
[New Thread 0x7fffc99f6700 (LWP 107599)]
Audio: frequency: 48000, channels: 2, samples: 256
sdl_create_buffers: creating stream buffer of 25600 bytes
Audio: End initialization
Keyboard: Start initialization
Input: Adding keyboard #0: System keyboard (device id: System keyboard)
Keyboard: Registered System keyboard
Keyboard: End initialization
Mouse: Start initialization
Input: Adding mouse #0: System mouse (device id: System mouse)
Mouse: Registered System mouse
Mouse: End initialization
Lightgun: Begin initialization
Lightgun: End initialization
Joystick mapping: Logical id 1: MicrosoftX-BoxOnepad(Firmware2015)
Joystick: Start initialization
Input: Adding joystick #0: MicrosoftX-BoxOnepad(Firmware2015) (device id: MicrosoftX-BoxOnepad(Firmware2015))
Joystick: 78081g503.ic655 NOT FOUND (NO GOOD DUMP KNOWN) (tried in tgmj coh3002c tgmj)
WARNING: the machine might not run correctly.

Thread 1 "mame" received signal SIGSEGV, Segmentation fault.
0x0000003b00584500 in ?? ()
Missing separate debuginfos, use: dnf debuginfo-install dbus-libs-1.12.16-3.fc31.x86_64 gsm-1.0.18-5.fc31.x86_64 libICE-1.0.10-2.fc31.x86_64 libSM-1.2.3-4.fc31.x86_64 libX11-xcb-1.6.8-3.fc31.x86_64 libXScrnSaver-1.2.3-4.fc31.x86_64 libXcursor-1.1.15-6.fc31.x86_64 libXfixes-5.0.3-10.fc31.x86_64 libXinerama-1.1.4-4.fc31.x86_64 libXrandr-1.5.2-2.fc31.x86_64 libXrender-0.9.10-10.fc31.x86_64 libXxf86vm-1.1.4-12.fc31.x86_64 libcap-2.26-6.fc31.x86_64 libdrm-2.4.100-1.fc31.x86_64 libuuid-2.34-3.fc31.x86_64
(gdb) bt
#0  0x0000003b00584500 in ?? ()
#1  0x000055555d95f819 in chain_manager::update_screen_textures ()
    at ../../../../../src/osd/modules/render/bgfx/chainmanager.cpp:451
#2  0x000055555d90457a in renderer_bgfx::draw () at ../../../../../src/osd/modules/render/drawbgfx.cpp:845
#3  0x000055555d8ddb18 in sdl_window_info::update () at ../../../../../src/osd/sdl/window.cpp:632
#4  sdl_window_info::update () at ../../../../../src/osd/sdl/window.cpp:560
#5  0x000055555d8db79b in sdl_osd_interface::update () at ../../../../../src/osd/sdl/video.cpp:108
#6  0x0000555560fc7a8f in video_manager::frame_update () at ../../../../../src/emu/video.cpp:237
#7  0x0000555560f8ab1a in screen_device::vblank_begin () at ../../../../../src/emu/screen.cpp:1597
#8  0x0000555560f8e92d in screen_device::device_timer () at ../../../../../src/emu/screen.cpp:895
#9  0x0000555560f852cf in device_t::timer_expired () at ../../../../../src/emu/device.h:608
#10 device_scheduler::execute_timers () at ../../../../../src/emu/schedule.cpp:910
#11 device_scheduler::timeslice () at ../../../../../src/emu/schedule.cpp:534
#12 0x0000555560f33140 in running_machine::run () at ../../../../../src/emu/machine.cpp:367
#13 0x000055555d9b4372 in mame_machine_manager::execute () at ../../../../../src/frontend/mame/mame.cpp:261
#14 0x000055555da4d027 in cli_frontend::start_execution () at ../../../../../src/frontend/mame/clifront.cpp:264
#15 0x000055555da4d282 in cli_frontend::execute () at ../../../../../src/frontend/mame/clifront.cpp:280
#16 0x000055555d9b291a in emulator_info::start_frontend () at ../../../../../src/frontend/mame/mame.cpp:392
#17 0x0000555559750c09 in main () at ../../../../../src/osd/sdl/sdlmain.cpp:217
(gdb)

**Comment by MooglyGuy** [comment]
Unfortunately, there's not a lot that I can do with just a callstack. I need to be able to examine local variables and such. If possible, can you please provide dumps of the relevant class members and function locals in update_screen_textures? Thanks!

**Comment by belegdol** [comment]
I will try later tonight. Would bt full output provide the dump you need?

**Comment by cuavas** [comment]
You'll need to catch it in the debugger, back up a frame, and then `p *this`.  You might have to rebuild with `SYMLEVEL=2` to get local variable symbols (be warned there's a big increase in executable size and build time - you might want to do a single-driver build).

**Comment by belegdol** [comment]
Ok, will try - wish me luck 😅

**Comment by cuavas** [comment]
Please build with 06f99c21a08fe76ae833a5587d19c33ed0a035f1 or later and re-test.  This fixed the tgmj crash on resolution change for me on Windows 10 Pro 1803 with GeForce MX130.

**Comment by belegdol** [comment]
I am pleased to report that with a970ef42f1111e5f3c5e5b2632ad5c73d6fb9fa4 the crashes on linux are gone. I have tested ikaruga, firefox, tgmj, umk3 and invqix and all can be started with -bgfx_screen_chains hlsl with both opengl and vulkan.
Windows was more crashy, I will test that one later.

**Comment by belegdol** [comment]
Windows 10 is looking much better too. Ikaruga only crashes with dx12, with both hlsl and default screen chain, and shows no video with dx9 (known issue).
With firefox:
- default screen chain: dx9 shows video with weird colours, while dx12 crashes on exit (also a known problem), other backends work.
- hlsl: there is only HUD shown with dx9, dx12 crashes on exit, other backends work

I would say this issue can be closed, I will file a separate one for ikaruga dx12 crash. Thank you both!
